In [2]:
import pandas as pd
from IPython.display import display, Markdown

In [3]:
ad_data = pd.read_csv("/windows/local-git-repos/future_ds_02/data.csv")

In [4]:
def Explore(df, title):
    display(Markdown(title))
    display(df.head(20))
    display(Markdown("---"))
    display(Markdown("### Shape"))
    display(df.shape)
    display(Markdown("---"))
    display(Markdown("### Missing Values"))
    display(df.isnull().sum().sort_values(ascending=False))    
    display(Markdown("---"))
    display(Markdown("### Data Types"))
    display(df.dtypes)     
    display(Markdown("---"))
    display(Markdown("### unique Values Per Column"))
    display(df.nunique())            

Explore(ad_data, "## Facebook Ad Campaign Data")    

## Facebook Ad Campaign Data

ad_id reporting_start reporting_end campaign_id fb_campaign_id    age  \
0   708746      17/08/2017    17/08/2017         916         103916  30-34   
1   708749      17/08/2017    17/08/2017         916         103917  30-34   
2   708771      17/08/2017    17/08/2017         916         103920  30-34   
3   708815      30/08/2017    30/08/2017         916         103928  30-34   
4   708818      17/08/2017    17/08/2017         916         103928  30-34   
5   708820      17/08/2017    17/08/2017         916         103929  30-34   
6   708889      17/08/2017    17/08/2017         916         103940  30-34   
7   708895      17/08/2017    17/08/2017         916         103941  30-34   
8   708953      17/08/2017    17/08/2017         916         103951  30-34   
9   708958      30/08/2017    30/08/2017         916         103952  30-34   
10  708979      30/08/2017    30/08/2017         916         103955  30-34   
11  709023      29/08/2017    29/08/2017         916         103962  30-34   
12  709038      29/08/2017    29/08/2017         916         103965  30-34   
13  709040      29/08/2017    29/08/2017         916         103965  30-34   
14  709059      30/08/2017    30/08/2017         916         103968  30-34   
15  709105      30/08/2017    30/08/2017         916         103976  30-34   
16  709115      30/08/2017    30/08/2017         916         103978  30-34   
17  709124      30/08/2017    30/08/2017         916         103979  30-34   
18  709179      30/08/2017    30/08/2017         916         103988  35-39   
19  709183      30/08/2017    30/08/2017         916         103989  35-39   

   gender  interest1  interest2  interest3  impressions  clicks  spent  \
0       M         15         17         17       7350.0       1   1.43   
1       M         16         19         21      17861.0       2   1.82   
2       M         20         25         22        693.0       0   0.00   
3       M         28         32         32       4259.0       1   1.25   
4       M         28         33         32       4133.0       1   1.29   
5       M         29         30         30       1915.0       0   0.00   
6       M         15         16         17      15615.0       3   4.77   
7       M         16         20         18      10951.0       1   1.27   
8       M         27         31         31       2355.0       1   1.50   
9       M         28         32         31       9502.0       3   3.16   
10      M         31         37         34       1224.0       0   0.00   
11      M          7          8          8        735.0       0   0.00   
12      M         16         20         22       5117.0       0   0.00   
13      M         16         19         21       5120.0       0   0.00   
14      M         20         26         23      14669.0       7  10.28   
15      M         28         30         32       1241.0       0   0.00   
16      M         30         34         31       2305.0       1   0.57   
17      M         31         34         33       1024.0       0   0.00   
18      M         15         16         17       4627.0       1   1.69   
19      M         16         20         22      21026.0       4   4.63   

    total_conversion  approved_conversion  
0                2.0                  1.0  
1                2.0                  0.0  
2                1.0                  0.0  
3                1.0                  0.0  
4                1.0                  1.0  
5                1.0                  1.0  
6                1.0                  0.0  
7                1.0                  1.0  
8                1.0                  0.0  
9                1.0                  0.0  
10               1.0                  0.0  
11               1.0                  0.0  
12               1.0                  0.0  
13               1.0                  0.0  
14               1.0                  1.0  
15               1.0                  1.0  
16               1.0                  0.0  
17               1.

---

### Shape

(1143, 15)

---

### Missing Values

total_conversion       382
approved_conversion    382
ad_id                    0
campaign_id              0
fb_campaign_id           0
reporting_start          0
reporting_end            0
gender                   0
age                      0
interest1                0
interest2                0
impressions              0
interest3                0
spent                    0
clicks                   0
dtype: int64

---

### Data Types

ad_id                    int64
reporting_start         object
reporting_end           object
campaign_id             object
fb_campaign_id          object
age                     object
gender                  object
interest1                int64
interest2                int64
interest3                int64
impressions            float64
clicks                   int64
spent                  float64
total_conversion       float64
approved_conversion    float64
dtype: object

---

### unique Values Per Column

ad_id                  1143
reporting_start          14
reporting_end            14
campaign_id               7
fb_campaign_id          490
age                      44
gender                   63
interest1                64
interest2               431
interest3               168
impressions            1126
clicks                   97
spent                   509
total_conversion         24
approved_conversion      15
dtype: int64

- ad_id: &nbsp;&nbsp;&nbsp; A unique identifier for each individual ad (creative or ad set) in the campaign. Each row corresponds to one ad instance.
- reporting_start: &nbsp;&nbsp;&nbsp; The start date of the reporting period — when Facebook started tracking the ad’s performance for that record.
- reporting_end: &nbsp;&nbsp;&nbsp; The end date of the reporting period — when Facebook stopped tracking for that record. Typically the same as reporting_start if it’s daily data.
- campaign_id: &nbsp;&nbsp;&nbsp; The internal campaign ID assigned by the advertiser or data provider. Several ads (ad_id) can belong to one campaign.
- fb_campaign_id: &nbsp;&nbsp;&nbsp; The Facebook-generated campaign ID — a unique identifier Facebook assigns for tracking across its system. It can differ from the advertiser’s own campaign_id.
- age: &nbsp;&nbsp;&nbsp; The age range of the audience segment targeted by the ad, e.g. "30-34". Used for demographic targeting.
- gender: &nbsp;&nbsp;&nbsp; The gender of the targeted audience (e.g., "M", "F"). Helps measure ad performance by gender.
- interests 1 - 3: &nbsp;&nbsp;&nbsp; Encoded numerical representations of Facebook audience interest categories (e.g., “fitness,” “technology,” etc.). These IDs represent topics the targeted users have shown interest in. Multiple columns imply combinations or layered targeting.
- impressions: &nbsp;&nbsp;&nbsp; The number of times the ad was displayed on users’ screens. One person can see the ad multiple times, so impressions ≥ reach.
- clicks: &nbsp;&nbsp;&nbsp; The number of times users clicked on the ad (e.g., clicked a link, call-to-action, or image). Used to calculate CTR (Click-Through Rate).
- spent: &nbsp;&nbsp;&nbsp; The amount of money (in currency units) spent on showing that ad during the reporting period.
- total_conversion: &nbsp;&nbsp;&nbsp; The total number of desired actions taken by users after interacting with the ad (e.g., purchases, signups, app installs). This can include both approved and unapproved conversions.
- approved_conversion: &nbsp;&nbsp;&nbsp; The number of conversions approved (validated by the advertiser or platform, e.g., confirmed purchases). Usually a subset of total_conversion.

In [5]:
ad_data['campaign_id'].value_counts()

campaign_id
936      464
1178     243
45-49    144
30-34     99
40-44     71
35-39     68
916       54
Name: count, dtype: int64

- Upon manual csv exploration after the above code snippet and some sql failures, I noticed missing values from row 763 onward to 1143 where campaign_id and fb_campaign_id were missing however all columns were accounted for by trailling commas at the end.
- The code below is designed to remove the trailling commas and leave the actual Null values in the respective columns, getting rid of alignment.

In [6]:
import csv

# Open 2 files using a context manager to ensure they are closed afterwards
with open('data.csv', 'r') as infile, open('data-fixed.csv', 'w', newline='') as outfile:

    # Create an iterator to read from the corrupted csv and to write to a fixed csv file
    reader = csv.reader(infile)
    writer = csv.writer(outfile)

    # Create the headers in the new csv file
    headers = next(reader)
    expected_columns = len(headers)
    writer.writerow(headers)

    for row_num, row in enumerate(reader, start=2):
        # If we have exactly 13 fields + 2 trailing empties = 15 fields but wrong positions
        # AND if position 3 contains an age pattern
        if len(row) == expected_columns and '-' in str(row[3]) and len(row[3]) <= 6:

            # Remove trailing empty values
            while row and row[-1] == '':
                row.pop()

            # Now insert empty strings at position 3 and 4
            row = row[:3] + ['', ''] + row[3:]

        writer.writerow(row)

print("CSV fixed! Saved as data_fixed.csv")


CSV fixed! Saved as data_fixed.csv


In [7]:
ad_data_fixed = pd.read_csv("/windows/local-git-repos/future_ds_02/data-fixed.csv")

Explore(ad_data_fixed, "## Facebook Ad Campaign Data Fixed")   

## Facebook Ad Campaign Data Fixed

ad_id reporting_start reporting_end  campaign_id  fb_campaign_id    age  \
0   708746      17/08/2017    17/08/2017        916.0        103916.0  30-34   
1   708749      17/08/2017    17/08/2017        916.0        103917.0  30-34   
2   708771      17/08/2017    17/08/2017        916.0        103920.0  30-34   
3   708815      30/08/2017    30/08/2017        916.0        103928.0  30-34   
4   708818      17/08/2017    17/08/2017        916.0        103928.0  30-34   
5   708820      17/08/2017    17/08/2017        916.0        103929.0  30-34   
6   708889      17/08/2017    17/08/2017        916.0        103940.0  30-34   
7   708895      17/08/2017    17/08/2017        916.0        103941.0  30-34   
8   708953      17/08/2017    17/08/2017        916.0        103951.0  30-34   
9   708958      30/08/2017    30/08/2017        916.0        103952.0  30-34   
10  708979      30/08/2017    30/08/2017        916.0        103955.0  30-34   
11  709023      29/08/2017    29/08/2017        916.0        103962.0  30-34   
12  709038      29/08/2017    29/08/2017        916.0        103965.0  30-34   
13  709040      29/08/2017    29/08/2017        916.0        103965.0  30-34   
14  709059      30/08/2017    30/08/2017        916.0        103968.0  30-34   
15  709105      30/08/2017    30/08/2017        916.0        103976.0  30-34   
16  709115      30/08/2017    30/08/2017        916.0        103978.0  30-34   
17  709124      30/08/2017    30/08/2017        916.0        103979.0  30-34   
18  709179      30/08/2017    30/08/2017        916.0        103988.0  35-39   
19  709183      30/08/2017    30/08/2017        916.0        103989.0  35-39   

   gender  interest1  interest2  interest3  impressions  clicks  spent  \
0       M         15         17         17         7350       1   1.43   
1       M         16         19         21        17861       2   1.82   
2       M         20         25         22          693       0   0.00   
3       M         28         32         32         4259       1   1.25   
4       M         28         33         32         4133       1   1.29   
5       M         29         30         30         1915       0   0.00   
6       M         15         16         17        15615       3   4.77   
7       M         16         20         18        10951       1   1.27   
8       M         27         31         31         2355       1   1.50   
9       M         28         32         31         9502       3   3.16   
10      M         31         37         34         1224       0   0.00   
11      M          7          8          8          735       0   0.00   
12      M         16         20         22         5117       0   0.00   
13      M         16         19         21         5120       0   0.00   
14      M         20         26         23        14669       7  10.28   
15      M         28         30         32         1241       0   0.00   
16      M         30         34         31         2305       1   0.57   
17      M         31         34         33         1024       0   0.00   
18      M         15         16         17         4627       1   1.69   
19      M         16         20         22        21026       4   4.63   

    total_conversion  approved_conversion  
0                  2                    1  
1                  2                    0  
2                  1                    0  
3                  1                    0  
4                  1                    1  
5                  1                    1  
6                  1                    0  
7                  1                    1  
8                  1                    0  
9                  1                    0  
10                 1                    0  
11                 1                    0  
12                 1                    0  
13                 1                    0  
14                 1                    1  
15                 1                    1  
16                 1 

---

### Shape

(1143, 15)

---

### Missing Values

campaign_id            382
fb_campaign_id         382
ad_id                    0
reporting_end            0
reporting_start          0
age                      0
gender                   0
interest1                0
interest2                0
interest3                0
impressions              0
clicks                   0
spent                    0
total_conversion         0
approved_conversion      0
dtype: int64

---

### Data Types

ad_id                    int64
reporting_start         object
reporting_end           object
campaign_id            float64
fb_campaign_id         float64
age                     object
gender                  object
interest1                int64
interest2                int64
interest3                int64
impressions              int64
clicks                   int64
spent                  float64
total_conversion         int64
approved_conversion      int64
dtype: object

---

### unique Values Per Column

ad_id                  1143
reporting_start          14
reporting_end            14
campaign_id               3
fb_campaign_id          488
age                       4
gender                    2
interest1                40
interest2                66
interest3                69
impressions            1130
clicks                  183
spent                   869
total_conversion         32
approved_conversion      16
dtype: int64